### PR TITLE
Introduce TermListPatcher and AliasGroupListPatcher

### PR DESCRIPTION
**Pull request by Benestar**
This pull request is mainly about moving code around and creating a lot of test cases.

It consists of three commits that can be reviewed individually.

After this pull request, FingerprintPatcher is unused and can be removed. We have to make sure that all test cases in FingerprintPatcherTest are also present in TermListPatcherTest and AliasGroupListPatcher test.

To make the new classes actually useful, we have to move them out of the Internal namespace and make them package public.

### CODE REVIEW

File: src/Diff/ItemPatcher.php [hunk 1]
```diff
@@ -18,9 +19,14 @@
 class ItemPatcher implements EntityPatcherStrategy {
```
Comment (thiemowmde) [inline src/Diff/ItemPatcher.php:18]: Uh, **this** is weird in my opinion. Item and Property are guaranteed to have a Fingerprint. We do not win anything by using the smaller patchers here.

File: src/Diff/PropertyPatcher.php [hunk 2]
```diff
@@ -17,9 +18,14 @@
 class PropertyPatcher implements EntityPatcherStrategy {
```
Comment (thiemowmde) [inline src/Diff/PropertyPatcher.php]: As above.

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+ */
+class AliasGroupListPatcherTest extends PHPUnit_Framework_TestCase {
+
+	public function providePatchAliasGroupList() {
+		return array(
+			'add aliases' => array(
+				new AliasGroupList(),
+				new Diff( array(
+					'en' => new Diff( array( new DiffOpAdd( 'foo' ), new DiffOpAdd( 'bar' ) ) )
+				) ),
```
Comment (thiemowmde) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php:49]: I notice there is never a true/false here, so all `Diff` instances rely on their "is associative" auto-detection. I would either avoid that or add tests for all three possible values true/false/null.

**Comment by Benestar** [comment]
I removed the questioned commit from the chain.

**Comment by thiemowmde** [comment]
PHPMD. :-(

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				new AliasGroupList( array(
+					new AliasGroup( 'en', array( 'foo', 'bar' ) )
+				) )
+			),
+			'change aliases is no-op' => array(
+				new AliasGroupList( array(
+					new AliasGroup( 'en', array( 'foo', 'bar' ) )
+				) ),
+				new Diff( array(
+					'en' => new Diff( array( new DiffOpChange( 'bar', 'baz' ) ) )
```
Comment (thiemowmde) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php:68]: This looks like a valid change. Why should this be a no-op? Especially since the very similar change below works.

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+					new AliasGroup( 'en', array( 'foo', 'bar', 'baz' ) )
+				) ),
+				new Diff( array(
+					'en' => new DiffOpRemove( array( 'foo', 'bar' ) )
+				) ),
+				new AliasGroupList( array(
+					new AliasGroup( 'en', array( 'foo', 'bar', 'baz' ) )
+				) )
+			),
+			'add an existing language is no-op' => array(
```
Comment (thiemowmde) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php]: The test is correct, but the description should not talk about a "no-op".

File: tests/unit/Diff/Internal/TermListPatcherTest.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+					new Term( 'en', 'foo' ),
+					new Term( 'de', 'bar' ),
+					new Term( 'nl', 'baz' )
+				) ),
+				new Diff( array(
+					'en' => new DiffOpChange( 'foo', 'bar' ),
+					'de' => new DiffOpRemove( 'bar' ),
+					'nl' => new DiffOpChange( 'bar', 'foo' ),
+					'it' => new DiffOpAdd( 'foo' )
+				) ),
```
Comment (thiemowmde) [inline tests/unit/Diff/Internal/TermListPatcherTest.php:161]: All `Diff` instances that contain language keys need to be marked as associative with `true`.

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+				new AliasGroupList( array(
+					new AliasGroup( 'en', array( 'foo', 'bar' ) )
+				) )
+			),
+			'change different aliases is no-op' => array(
+				new AliasGroupList( array(
+					new AliasGroup( 'en', array( 'foo' ) )
+				) ),
+				new Diff( array(
+					'en' => new Diff( array( new DiffOpChange( 'bar', 'baz' ) ) )
```
Comment (thiemowmde) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php:202]: All test cases that are nested like this (`Diff` → `Diff` → `DiffOpChange`) should be duplicated, one with `Diff` → `Diff( …, true )` → `DiffOpChange` and one with `Diff` → `Diff( …, false )` → `DiffOpChange`.

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php (hunk 5 above)
Comment (Benestar) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php]: C&p mistake, will fix

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php (hunk 4 above)
Comment (Benestar) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php:68]: If I understand it correctly, that is basically what https://github.com/wmde/Diff/pull/65 is about. I just wanted to test the current behaviour but we should change this test if the other pull request gets merged.

File: tests/unit/Diff/Internal/TermListPatcherTest.php (hunk 6 above)
Comment (Benestar) [inline tests/unit/Diff/Internal/TermListPatcherTest.php:161]: How does this still work without explicitely stating that it is associative?

**Comment by Benestar** [comment]
> I notice there is never a true/false here, so all Diff instances rely on their "is associative" auto-detection. I would either avoid that or add tests for all three possible values true/false/null.

I didn't yet completely understand what the $isAssociative is about. Is it just the array of diffs that can be associative or linear? Anyways, it would perhaps be nice to do that in the actual test so that we don't have too much duplication.

> PHPMD. :-(

Any idea how to fix that? Is it possible to disable the linting per file/method?

File: tests/unit/Diff/Internal/TermListPatcherTest.php (hunk 6 above)
Comment (JeroenDeDauw) [inline tests/unit/Diff/Internal/TermListPatcherTest.php:161]: When not stated explicitly, it is determined by seeing if there are changed that are no add or remove.

**Comment by JeroenDeDauw** [comment]
> I didn't yet completely understand what the $isAssociative is about. Is it just the array of diffs that can be associative or linear?

The diffs as well. See the first 3 sections in "usage" at https://github.com/wmde/Diff#usage

**Comment by JeroenDeDauw** [comment]
> To make the new classes actually useful, we have to move them out of the Internal namespace and make them package public.

Sounds good. I'm guessing this was not done so far as we actually had no need to expose those things. I don't actually recall the reason though.

**Comment by thiemowmde** [comment]
Rebased. Note that this patch, as it is now, reverts #120! I suggest we fix #131 first before continuing to work on this.

File: tests/unit/Diff/Internal/AliasGroupListPatcherTest.php (hunk 4 above)
Comment (thiemowmde) [inline tests/unit/Diff/Internal/AliasGroupListPatcherTest.php:68]: Fixed in #131.

**Comment by thiemowmde** [comment]
Rebased again, on top of #120 and #131. All this patch now does is splitting the existing implementation (now without reverting anything) and adding tests. The only relevant detail I had to change in the tests was the one #131 fixed (see https://github.com/wmde/WikibaseDataModelServices/pull/125#discussion_r61218037). I also added a few more test cases with the "associative" flag set to true and false.

From my point of view this is ready to be merged.

Open questions (for other patches, should not block this one):
- Which classes must be public?
- Can we fix and use https://github.com/wmde/Diff/pull/65 before releasing this?

**Comment by adrianheine** [comment]
Needs manual rebase
